metpy/plots/declarative.py

```diff
@@ -0,0 +1,423 @@
+
+        Resets all internal state and sets need for redraw.
+
+        """
+        if getattr(self, 'handle', None) is not None:
+            self.clear_handle()
+            self._need_redraw = True
+
+    def clear_handle(self):
+        """Clear the handle to the plot instance."""
+        self.handle.remove()
+        self.handle = None
+
+    @observe('parent')
+    def _parent_changed(self, _):
+        """Handle setting the parent object for the plot."""
+        self.clear()
+
+    @observe('field', 'level', 'time')
+    def _update_data(self, _=None):
+        """Handle updating the internal cache of data.
+
+        Responds to changes in various subsetting parameters.
+
+        """
+        self._griddata = None
+        self.clear()
+
+    # Can't be a Traitlet because notifications don't work with arrays for traits
+    # notification never happens
+    @property
+    def data(self):
+        """Access the current data subset."""
+        return self._data
+
+    @data.setter
+    def data(self, val):
+        self._data = val
+        self._update_data()
+
+    @property
+    def griddata(self):
+        """Return the internal cached data."""
+        if getattr(self, '_griddata', None) is None:
+
+            if self.field:
+                data = self.data.metpy.parse_cf(self.field)
+            elif not hasattr(self.data.metpy, 'x'):
+                # Handles the case where we have a dataset but no specified field
+                raise ValueError('field attribute has not been set.')
+            else:
+                data = self.data
+
+            subset = {'method': 'nearest'}
+            if self.level is not None:
+                subset[data.metpy.vertical.name] = self.level
+
+            if self.time is not None:
+                subset[data.metpy.time.name] = self.time
+            self._griddata = data.metpy.sel(**subset).squeeze()
+
+        return self._griddata
+
+    @property
+    def plotdata(self):
+        """Return the data for plotting.
+
+        The data array, x coordinates, and y coordinates.
+
+        """
+        x = self.griddata.metpy.x
+        y = self.griddata.metpy.y
+
+        if 'degree' in x.units:
+            import numpy as np
+            x, y, _ = self.griddata.metpy.cartopy_crs.transform_points(ccrs.PlateCarree(),
+                                                                       *np.meshgrid(x, y)).T
+            x = x[:, 0] % 360
+            y = y[0, :]
+
+        return x, y, self.griddata
+
+    @property
+    def name(self):
+        """Generate a name for the plot."""
+        ret = self.field
+        if self.level is not None:
+            ret += '@{:d}'.format(self.level)
+        return ret
+
+    def draw(self):
+        """Draw the plot."""
+        if self._need_redraw:
+            if getattr(self, 'handle', None) is None:
+                self._build()
+            self._need_redraw = False
+
+
+@exporter.export
+class ImagePlot(Plot2D):
```

Comment:
I'm not convinced this will ever be something we really want to support, but we can revisit in the future if a compelling use case appears. I don't think we're doing anything that would make it difficult to support (heck, it might actually work currently).

Reply:
I may be biased, but one good use case could be cross sections! Or, would that work better as a separate class?

Reply:
Darn you and your good ideas @jthielen 😉

Will need to wait for next release, but that's a really good one. We'll need a new panel type for that as well.